web/src/conversation_participants.ts

```diff
@@ -0,0 +1,46 @@
+import type {Message} from "./message_store.ts";
+import * as message_store from "./message_store.ts";
+import * as people from "./people.ts";
+
+export class ConversationParticipants {
+    humans: Set<number>;
+    bots: Set<number>;
+
+    constructor(messages: Message[]) {
+        this.humans = new Set();
+        this.bots = new Set();
+        this.add_from_messages(messages);
+    }
+
+    add_from_messages(messages: Message[]): void {
+        for (const msg of messages) {
+            if (msg.sent_by_me) {
+                this.humans.add(msg.sender_id);
+                continue;
+            }
+
+            const sender = people.maybe_get_user_by_id(msg.sender_id);
+            if (!sender) {
+                continue;
+            }
+
+            if (sender.is_bot) {
+                this.bots.add(msg.sender_id);
+            } else {
+                this.humans.add(msg.sender_id);
+            }
+        }
+    }
+
+    remove_from_message_ids(message_ids: number[]): void {
+        for (const message_id of message_ids) {
+            const msg = message_store.get(message_id);
+            if (!msg) {
+                continue;
+            }
+
+            this.humans.delete(msg.sender_id);
+            this.bots.delete(msg.sender_id);
```

Comment:
This algorithm seems wrong; if we remove a single message with a given sender, it drops that sender from the participants list entirely.

I think it's probably OK to just have this support recomputing the whole data structure from scratch to handle removals for now; it's a very fast loop.

Reply:
Oops! Changed to recompute the whole thing.